Build: Simplify Dashboard Build Toolchain

ockham opened the pull request:
WIP. Prep work for #12072.

One core thing I've identified is that `tools/builder/react.js` uses `static.jsx` to produce minimal static markup (HTML) versions of the dashboard (really just "Turn on your JavaScript" notices). It `require()`s the webpack-bundled version of `static.jsx` to that end, and attaches `renderToStaticMarkup()`-produced strings to `window`, which it then writes to HTML files.

Getting rid of the complexity added by passing information by attaching to `window` (and thus requiring `jsdom` etc) is the main objective of this PR.

Maybe we can instead produce those HTML files as by-products of the bundling, i.e. through a Webpack plugin or loader? https://github.com/GoogleChromeLabs/prerender-loader#string-prerendering looks very promising!

Alternatively, we shouldn't bundle but just transpile `static.jsx` so we can simply `require()` it from `tools/builder/react.js`.

#### Changes proposed in this Pull Request:

Don't use `window`/`jsdom` in build system.

#### Testing instructions:

```
yarn distclean && yarn
yarn build-client
yarn docker:up
```

- Navigate to `/wp-admin/admin.php?page=jetpack_modules`
- Disable JavaScript
- See this message?

![image](https://user-images.githubusercontent.com/96308/56582883-19c3c300-65fa-11e9-92d8-e3355f73317d.png)

#### Proposed changelog entry for your changes:

Build: Simplify Dashboard Build Toolchain

ockham committed:
Static.jsx: Named import

ockham committed:
Simplify mapDispatchToProps

ockham committed:
Doesn't need to be an async action

ockham committed:
Add prerender-loader devDep

ockham committed:
Add HtmlWebpackPlugin devDep

ockham committed:
First try

ockham commented on _inc/client/state/initial-state/actions.js:
```diff
@@ -3,11 +3,7 @@
-};
+export const setInitialState = () => ( {
+	type: JETPACK_SET_INITIAL_STATE,
+	initialState: window.Initial_State,
```
We should probably parametrize the action for `initialState` to put the burden of instantiating it with `window.Initial_State` on the consumer.

ockham commented on _inc/client/static.jsx:
```diff
@@ -12,23 +12,24 @@ import store from 'state/redux-store';
+				<StaticMain />
+			</Provider>
+		</div>
+	);
```
See https://github.com/GoogleChromeLabs/prerender-loader#string-prerendering

ockham commented on _inc/client/static.jsx:
```diff
@@ -12,23 +12,24 @@ import store from 'state/redux-store';
 );
 
-window.versionNotice = Server.renderToStaticMarkup(
+window.versionNotice = renderToStaticMarkup(
```
Will probably need to move `noscriptNotice` and `versionNotice` to separate files,

ockham committed:
Have setInitialState accept an initialState arg

ockham committed:
StaticMain: Set initialState based on const rather than window global

ockham committed:
Don't set individual reducers' initialState based on window global

ockham committed:
Use constructor() instead of UNSAFE_componentWillMount()

ockham committed:
Separate templates

ockham committed:
React Builder: Remove buildStatic task

ockham committed:
Kinda works

simison commented on tools/builder/react.js:
```diff
@@ -52,7 +48,7 @@ gulp.task( 'react:master', function( done ) {
 			}
 
-			buildStatic( done );
+			done();
```
Prolly just `onBuild.bind( this, done )` would work here but good also as-is.

ockham replied:
3a2ec1bba

simison commented:
Rebased

ockham committed:
Simplify onBuild

ockham committed:
Try StaticSiteGeneratorPlugin based approach

ockham committed:
State: Guard 2 Selectors in get()s

ockham committed:
Analytics: Check if window exists

ockham committed:
Actually render static.jsx

ockham committed:
Consolidate

ockham committed:
Remove now-obsolete dependencies

ockham committed:
TEMP: Comment out WordPressExternalDependenciesPlugin

ockham committed:
Remove obsolete settings

ockham committed:
Tools: Remove webpack.LoaderOptionsPlugin

This was solely meant to facilitate migration from Webpack v1 to v2. The debug option specifically was dropped by Webpack 3.

https://webpack.js.org/plugins/loader-options-plugin/

ockham committed:
Separate Webpack configs

ockham committed:
More safeguards

ockham commented:
Closing in favor of #12381.
